--- com/callassoftware/sample/pdfEngine/pdfToolboxSample.java.orig	2025-08-05 17:44:23.285256556 +0200
+++ com/callassoftware/sample/pdfEngine/pdfToolboxSample.java	2025-08-05 18:05:59.450571185 +0200
@@ -52,10 +52,21 @@
 			}
 
 			// init library
-			lib.init( args.get(0), args.get(1), null, "en");
+			lib.init( null, null, null, "en");
 			isInitialized = true;
 			System.out.println( "lib is initialized" );
 
+			String env_ls_urls = System.getenv("CALLAS_LICENSESERVER_URLS");
+			if (env_ls_urls == null) {
+				System.err.println("env var CALLAS_LICENSESERVER_URLS is empty - good bye");
+				System.exit(1); // Exit with error
+			}
+
+			String[] ls_urls = env_ls_urls.split(";");
+			String ls_msg = System.getenv("CALLAS_LICENSESERVER_MSG");
+			int ls_timeout=5;
+			int ls_types=(int)LicenseServerType.RUNCOUNT | LicenseServerType.CREDITS;
+			lib.licenseServerSetup(ls_urls, ls_timeout, ls_types, ls_msg);
 			// execute samples with args
 			lib.samples( args );
 		}
